### PR TITLE
Repair buffer leak when replying to NBNS requests

### PR DESCRIPTION
**Pull request by htibosch**
Description
-----------
As reported by [niconrok](https://forums.freertos.org/u/niconrok) in [this post](https://forums.freertos.org/t/memory-leak-during-nbns-llmnr-hostname-resolution/12062) the following:

when the application gives a positive reply to an NBNS requests, and only in case `BufferAllocation_2.c` is used, a network buffer will not be released.

The patch goes like this:
~~~c
 usLength = ( uint16_t ) ( sizeof( NBNSAnswer_t ) + ( size_t ) offsetof( NBNSRequest_t, usType ) );

 prvReplyDNSMessage( pxNetworkBuffer, ( BaseType_t ) usLength );

+if( pxNewBuffer != NULL )
+{
+    vReleaseNetworkBufferAndDescriptor( pxNewBuffer );
+}
~~~

Test Steps
-----------
Enabled NBNS and disable LLMNR:

    #define ipconfigUSE_NBNS   1
    #define ipconfigUSE_LLMNR  0

Write a version of `xApplicationDNSQueryHook()` that always returns `pdPASS`. In that case, the application will reply to all NBNS requests.
Issue some logging about the free network buffers ( `uxGetNumberOfFreeNetworkBuffers()` ).

Note that the LLMNR driver did not have this problem.

Related Issue
-----------
Both protocols can be very useful on a LAN: it allows to work with device names in stead of IP-addresses.
However, both protocols are not considered safe any more. There will be re better replacement: mDNS.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by AniruddhaKanhere — APPROVED** [review]

File: FreeRTOS_DNS.c [hunk 1]
```diff
@@ -1871,6 +1871,11 @@
                         usLength = ( uint16_t ) ( sizeof( NBNSAnswer_t ) + ( size_t ) offsetof( NBNSRequest_t, usType ) );
 
                         prvReplyDNSMessage( pxNetworkBuffer, ( BaseType_t ) usLength );
+
+                        if( pxNewBuffer != NULL )
```
Comment (muneebahmed10) [inline FreeRTOS_DNS.c:1875]: Reading the most recent [post](https://forums.freertos.org/t/memory-leak-during-nbns-llmnr-hostname-resolution/12062/7) it looks like `pxNewBuffer` is out of scope by this time? Line 1823 would need to be moved up.

Comment (AniruddhaKanhere) [reply FreeRTOS_DNS.c:1875]: I agree! Missed that completely.
@htibosch, would you mind making the requested changes?

**Review by AniruddhaKanhere — CHANGES_REQUESTED** [review]
See the comments.

File: FreeRTOS_DNS.c (hunk 1 above)
Comment (htibosch) [reply FreeRTOS_DNS.c:1875]: Sorry yes, you are right. I moved the declaration of `pxNewBuffer` upward and initialised it to NULL. Thanks

**Comment by htibosch** [comment]
About my commit "Variable 'pxNewBuffer' was out of scope" :

I moved the declaration of `pxNewBuffer` upward so that it can be released.

**Review by AniruddhaKanhere — APPROVED** [review]
@htibosch, thank you for the changes.

**Review by cobusve — APPROVED** [review]

**Pull request merged**